src/appleseed/renderer/modeling/phasefunction/henyeyphasefunction.cpp

```diff
@@ -0,0 +1,359 @@
+        if (std::abs(g) < +1e-5f)
+        {
+            cosine = s; // isotropic
+        }
+        else
+        {
+            const float t = (1.0f - sqr_g) / (1.0f + g*s);
+            cosine = 0.5f / g * (1.0f + sqr_g - t * t);
+        }
+        const float sine = std::sqrt(1.0f - cosine * cosine);
+
+        const Vector2f tangent =
+            sample_circle_uniform(sampling_context.next2<float>());
+
+        incoming =
+            basis.get_tangent_u() * tangent.x * sine +
+            basis.get_tangent_v() * tangent.y * sine +
+            basis.get_normal()    * cosine;
+
+        assert(feq(norm(incoming), 1.0f));
+
+        // Evaluate PDF.
+
+        const float numerator = (1.0f - sqr_g);
+        const float denominator = std::pow(1.0f + sqr_g - 2.0f*g*cosine, -1.5f);
+
+        return RcpFourPi<float>() * numerator * denominator;
+    }
+
+    virtual float evaluate(
+        const ShadingRay&     volume_ray,
+        const void*           data,
+        float                 distance,
+        const Vector3f&       incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const Vector3f outgoing = Vector3f(normalize(volume_ray.m_dir));
+        const float cosine = dot(incoming, outgoing);
+
+        //
+        // p(x) = 1/2 * (1 - g^2) / (1 + g^2 - 2gx)^(3/2),
+        // where x is cos(phi) and g is the average cosine parameter.
+        //
+
+        const float g = values->m_average_cosine;
+        const float sqr_g = g * g;
+
+        const float numerator = (1.0f - sqr_g);
+        const float denominator = std::pow(1.0f + sqr_g - 2.0f*g*cosine, -1.5f);
+
+        // Additionally divide by TwoPi, because we sample over the sphere.
+        return RcpFourPi<float>() * numerator * denominator;
+    }
+
+    virtual void evaluate_transmission(
+        const ShadingRay&     volume_ray,
+        const void*           data,
+        float                 distance,
+        Spectrum&             spectrum) const APPLESEED_OVERRIDE
+    {
+        extinction_coefficient(volume_ray, data, distance, spectrum);
+        spectrum = exp(-distance * spectrum);
+    }
+
+    virtual void evaluate_transmission(
+        const ShadingRay&     volume_ray,
+        const void*           data,
+        Spectrum&             spectrum) const APPLESEED_OVERRIDE
+    {
+        const float distance = static_cast<float>(
+            norm(volume_ray.m_dir) * (volume_ray.m_tmax - volume_ray.m_tmin));
+        evaluate_transmission(volume_ray, data, distance, spectrum);
+    }
+
+    virtual void scattering_coefficient(
+        const ShadingRay&     volume_ray,
+        const void*           data,
+        float                 distance,
+        Spectrum&             spectrum) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        spectrum = values->m_scattering * values->m_scattering_multiplier;
```

Comment:
A simple way to avoid one spectrum copy:

    spectrum = values->m_scattering;
    spectrum *= values->m_scattering_multiplier;